server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -607,6 +605,10 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                     }
                 });
             }, listener::onFailure);
+            // fork handling the above listener to the generic pool since it loads various pieces of metadata from the repository over a
+            // longer period of time
+            repository.getRepositoryData(new ThreadedActionListener<>(logger, clusterService.getClusterApplierService().threadPool(),
```

Comment:
Do we need to fork the listener handling? aren't we executing the get repository logic in `GENERIC` already?

Reply:
Due to the deduplication we don't have any guarantees as to what thread this will run on in the extremely unlikely but not impossible corner case where we run deduplication on deserialization from cache, there it could be any thread resolving the listener. 
=> we need to fork off here => I dropped the `GENERIC` on the upstream transport handler because we're forking here anyway.